### PR TITLE
sql: dynamically determine histogram sample size

### LINKED ISSUE #97701: opt: histograms for large tables can miss frequent values
By default we keep up to 10k samples during table statistics collection. This fixed sample size can be problematic for large tables. The chance that we will miss a frequent value (say a value used in 10k+ rows) increases as tables get larger (todo: plot this for tables with zipf-distributed data). We have seen at least one customer case with a histogram for a 600m+ row table missing a 30k+ value, so this is definitely possible in practice.

A couple ideas for how to address this:
- make sample size proportional to table size (probably means spilling to disk during table statistics collection)
- some sort of "lossless" histogram algorithm, since we're already doing a full scan of the table
- use a [confidence interval to estimate](https://en.wikipedia.org/wiki/Population_proportion#Estimation) row counts instead of calculating the expected value of the row count (similar to discussion in [#64570](https://github.com/cockroachdb/cockroach/issues/64570)) (but does this work when there are zero samples matching the constraint? the rule of thumb is that normality needs at least 10 observations)
- collect higher-resolution samples on known bad parts of the histogram (as described in [partial statistics collection](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20220126_partial_statistics_collection.md))

Jira issue: CRDB-24833
Epic: CRDB-34173

gz#19609

### LINKED ISSUE #123972: sql: dynamically determine histogram sample size
To collect histograms in table statistics, we currently sample 10k rows. 

The number of rows sampled when building histograms for table statistics is 10k, and it can be configured with [`sql.stats.histogram_samples.count`](https://www.cockroachlabs.com/docs/stable/cluster-settings#setting-sql-stats-histogram-samples-count). 

For very large table, this is too few rows. We should find a way to dynamically adjust the sample size based on the size of the table.

See this discussion for more context: https://cockroachlabs.slack.com/archives/C01RX2G8LT1/p1701834921110179?thread_ts=1701825005.623189&cid=C01RX2G8LT1

Jira issue: CRDB-38633

### PR DESCRIPTION
**Pull request by Uzair5162**
We currently sample a default of 10k rows when collecting stats to construct histograms. We have seen cases of histograms for large tables missing frequent values as a result of poor samples, and users have frequently reported slow queries as a result of this.

This commit dynamically picks a sample size based on estimated table size $n$, using a sample size of $582n^{0.29}$ bounded between 10k and 300k. This formula was derived using empirical research on the performance of different sample sizes using the code in effc5061c249a71ca0b4280052b29f97637eb50c. The formula approximates the following table:

| Table Size    | Sample Size |
| ------------- | ----------- |
| 10,000        | 10,000      |
| 100,000       | 15,000      |
| 1,000,000     | 30,000      |
| 10,000,000    | 60,000      |
| 100,000,000   | 100,000     |
| 1,000,000,000 | 300,000     |

These sample sizes empirically achieved the following coverage:

- 100k rows/15k samples: ~100% coverage of multiplicities down to 100x, ~80% down to 10x
- 1m rows/30k samples: ~100% coverage of multiplicities down to 1000x, ~95% down to 100x
- 10m rows/60k samples: ~100% coverage of multiplicities down to 10000x, ~95% down to 1000x, ~50% down to 100x
- 100m rows/100k samples: ~100% coverage of multiplicities down to 10000x, ~65% down to 1000x, ~10% down to 100x
- 1b rows/300k samples: ~100% coverage of multiplicities down to 100000x, ~95% down to 10000x, ~25% down to 1000x

Fixes: #123972

Fixes: #97701

Release note (sql change): Histograms are no longer constructed using a default sample size of 10k. Samples are now dynamically sized based on table size unless the sample count has been set in the table or cluster settings.

### CODE REVIEW

**Review by DrewKimball — COMMENTED** [review]
This is great work! The logic LGTM - I just have a few suggestions for tests.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-:-NzxJwu80AMN8BTQ-XuA:bd88ydl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Uzair5162)

___
*[`pkg/sql/distsql_plan_stats.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-NzxHXYj7tjQ6xgc8ziP:-NzxHXYj7tjQ6xgc8ziQ:bvwzf4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/60aee2552d2a7c3d539811c92e617e4cdc5ee691/pkg/sql/distsql_plan_stats.go#L88)):*
> ```Go
> 	)
> 	return uint32(numSamples)
> }
> ```

Can you add a unit test for this function? It would be nice to validate that all the cases mentioned in the commit message work correctly, as well as any "interesting" input values, you can think of, like a very large (think MaxUint64) row count.
___
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 210 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-NzxJJ9cBNdONp147wsc:-NzxJJ9cBNdONp147wsd:b-t7hjsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/60aee2552d2a7c3d539811c92e617e4cdc5ee691/pkg/sql/logictest/testdata/logic_test/distsql_stats#L210)):*
> ```
> 
> statement ok
> SET CLUSTER SETTING sql.stats.histogram_samples.count = 20000
> ```

Another test that might be interesting: inject some statistics with interesting values (like very large row count), and see if that causes any issues with the sample size calculation. There are a bunch of examples of statistics injection here: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/exec/execbuilder/testdata/forecast

**Review by Uzair5162 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125345)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/distsql_plan_stats.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-NzxHXYj7tjQ6xgc8ziP:-O-2SsDQ-u9MpewRRCM5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60aee2552d2a7c3d539811c92e617e4cdc5ee691/pkg/sql/distsql_plan_stats.go#L88)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can you add a unit test for this function? It would be nice to validate that all the cases mentioned in the commit message work correctly, as well as any "interesting" input values, you can think of, like a very large (think MaxUint64) row count.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 210 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-NzxJJ9cBNdONp147wsc:-O-2StlQBElNJTQyK-rZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60aee2552d2a7c3d539811c92e617e4cdc5ee691/pkg/sql/logictest/testdata/logic_test/distsql_stats#L210)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Another test that might be interesting: inject some statistics with interesting values (like very large row count), and see if that causes any issues with the sample size calculation. There are a bunch of examples of statistics injection here: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/exec/execbuilder/testdata/forecast
</blockquote></details>

Done.

File: pkg/sql/distsql_plan_stats.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ var histogramSamples = settings.RegisterIntSetting(
 	settings.ApplicationLevel,
 	"sql.stats.histogram_samples.count",
 	"number of rows sampled for histogram construction during table statistics collection",
```
Comment (rytaft) [inline pkg/sql/distsql_plan_stats.go:56]: nit: explain that 0 means this is set automatically

**Review by rytaft — APPROVED** [review]
Nice work!

File: pkg/sql/distsql_plan_stats.go [hunk 2]
```diff
@@ -71,6 +71,22 @@ var maxTimestampAge = settings.RegisterDurationSetting(
 	5*time.Minute,
 )
 
+// Dynamically determine the number of samples to collect based on the estimated
+// number of rows in the table. This formula is based on empirical data collected
+// by running the sampler with different sample sizes on a variety of
+// table sizes and observing the proportion of heavy hitters (most frequent
+// elements) represented in the sample.
```
Comment (rytaft) [inline pkg/sql/distsql_plan_stats.go:78]: nit: update the comment to mention that the returned value is between 10k and 300k

**Review by DrewKimball — APPROVED** [review]
:LGTM:

> Reviewed 2 of 2 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-:-O-4pcVE49-vZGkqOYvb:bmxilld)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @Uzair5162)

___
*[`pkg/sql/distsql_plan_stats_test.go` line 41 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-O-4oVdh26JJHJKc6Ktv:-O-4oVdh26JJHJKc6Ktw:b-y4r1l9) ([raw file](https://github.com/cockroachdb/cockroach/blob/43e761555adb13b1f48bc9f02be523e04e079284/pkg/sql/distsql_plan_stats_test.go#L41)):*
> ```Go
> 	}
> 
> 	t.Logf("max int: %d", math.MaxInt)
> ```

[nit] Did you mean to leave this in?

**Review by Uzair5162 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125345)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[`pkg/sql/distsql_plan_stats_test.go` line 41 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-O-4oVdh26JJHJKc6Ktv:-O-6qO113s3PtcU2pnEN:blr6pm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/43e761555adb13b1f48bc9f02be523e04e079284/pkg/sql/distsql_plan_stats_test.go#L41)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Did you mean to leave this in?
</blockquote></details>

oops, good catch!

**Review by Uzair5162 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125345)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[`pkg/sql/distsql_plan_stats.go` line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125345#gh-1633728667:-O-6uyhG6XhhqzfI4Fs7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/649137717e1718ebc86088a02c0643a28fce603c/pkg/sql/distsql_plan_stats.go#L56)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: explain that 0 means this is set automatically 
</blockquote></details>

Done.
___
*[`pkg/sql/distsql_plan_stats.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125345#gh-1633731083:-O-6v0EDEVQKM4RhSPpS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/649137717e1718ebc86088a02c0643a28fce603c/pkg/sql/distsql_plan_stats.go#L78)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: update the comment to mention that the returned value is between 10k and 300k
</blockquote></details>

Done.

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125345#-:-O-7ARxx4e3kx5U1k2O3:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

**Comment by mgartner** [comment]
Nice work! :lgtm:

**Pull request merged**